src/main/java/com/faforever/client/replay/ReplayService.java

```diff
@@ -373,6 +401,94 @@ private void runFafReplayFile(Path path) throws IOException, CompressorException
     gameService.runWithReplay(tempSupComReplayFile, replayId, gameType, version, modVersions, simMods, mapName);
   }
 
+  private void runRehostOnlineReplay(int replayId) {
+    downloadReplay(replayId)
+        .thenAccept((path) -> {
+          try {
+            runRehostFile(path);
+          } catch (IOException | CompressorException e) {
+            throw new RuntimeException(e);
+          }
+        })
+        .exceptionally(throwable -> {
+          if (throwable.getCause() instanceof FileNotFoundException) {
+            log.warn("Replay not available on server yet", throwable);
+            notificationService.addImmediateWarnNotification("replayNotAvailable", replayId);
+          } else {
+            log.error("Replay could not be started", throwable);
+            notificationService.addImmediateErrorNotification(throwable, "replayCouldNotBeStarted", replayId);
+          }
+          return null;
+        });
+  }
+
+  private void runRehostFafFile(Path path) throws IOException, CompressorException {
+    ReplayDataParser replayData = replayFileReader.parseReplay(path);
+    byte[] rawReplayBytes = replayData.getData();
+
+    Path tempSupComReplayFile = preferencesService.getPreferences().getData().getCacheDirectory().resolve(TEMP_SCFA_REPLAY_FILE_NAME);
+
+    Files.createDirectories(tempSupComReplayFile.getParent());
+    Files.copy(new ByteArrayInputStream(rawReplayBytes), tempSupComReplayFile, StandardCopyOption.REPLACE_EXISTING);
+
+    ReplayMetadata replayMetadata = replayData.getMetadata();
+    String gameType = replayMetadata.getFeaturedMod();
+    Integer replayId = replayMetadata.getUid();
+    Map<String, Integer> modVersions = replayMetadata.getFeaturedModVersions();
+    String mapName = parseMapFolderName(replayData);
+
+    Set<String> simMods = parseModUIDs(replayData);
+
+    Integer version = parseSupComVersion(replayData);
+    FeaturedModBean featuredMod;
+    try {
+      featuredMod = modService.getFeaturedMod(gameType).get();
+      NewGameInfo newGameInfo = new NewGameInfo(
+          replayMetadata.getTitle(),
+          Strings.emptyToNull(""),
+          featuredMod,
+          mapName,
+          simMods,
+          GameVisibility.PUBLIC,
+          null,
+          null,
+          false);
+
+      gameService.hostGame(newGameInfo);
```

Comment:
This all should be moved to the game service. Additionally don't make it blocking with the get function rather it should return the chained completable future.

Reply:
Actually scratch this. Looking at the current pattern we have it is fine to have this in replayService. However it should be made non blocking, by chaining the futures. Also it would be nice to inherit the things last rating limits and password from the last hosted game

Reply:
I specifically didn't bring over rating limits etc because if I'm hosting from someone else's replay I probably don't have the same requirements - as it happens I don't see the password or rating limits in the replay data, so it looks like I couldn't get that anyway.

Reply:
Can we consider this a follow up to this feature perhaps? 
It opens questions like, what if the player hasn't hosted a game yet during this play session (I assume use the current default), but also what if they have played a game but don't want the password and rating limitations to apply to this hosting session? Should we offer UI to set password and rating limits when hosting from this gesture? (I'd prefer not as its extra friction for the user, but there's probably a conversation there).